libdevcore/MemoryDB.cpp

```diff
@@ -88,17 +106,17 @@ void MemoryDB::commit(std::unique_ptr<WriteBatchFace> _batch)
 // to pass in a function `f`, which will be called with the key and value
 // of each record in the database. If `f` returns false, the `forEach`
 // method must return immediately.
-void MemoryDB::forEach(std::function<bool(Slice, Slice)> f) const
+void MemoryDB::forEach(std::function<bool(Slice, Slice)> _f) const
 {
-    std::lock_guard<std::mutex> lock(m_mutex);
-    for (const auto& e : m_db)
+    dev::Guard lock(m_mutex);
+    for (auto const& e : m_db)
     {
-        if (!f(Slice(e.first.c_str()), Slice(e.second.c_str())))
+        if (!_f(Slice(e.first.c_str()), Slice(e.second.c_str())))
```

Comment:
Constructing `Slice` from `c_str()` also assumes it's zero-terminated string, which it is not. Look at other `vector_ref` constructors, is there constructor just from `std::string` ?